src/runtime/rust/src/runtime/array.rs

```diff
@@ -0,0 +1,393 @@
+use std::{
+  any::TypeId,
+  convert::TryFrom,
+  mem,
+  os::raw::{c_int, c_void},
+  ptr,
+  slice,
+};
+
+use ndarray;
+
+use super::allocator::Allocation;
+use errors::*;
+use ffi::runtime::{
+  DLContext, DLDataType, DLDataTypeCode_kDLFloat, DLDataTypeCode_kDLInt, DLDataTypeCode_kDLUInt,
+  DLDeviceType_kDLCPU, DLTensor,
+};
+
+#[derive(PartialEq)]
+pub enum Storage<'a> {
+  Owned(Allocation),
+  View(&'a mut [u8], usize), // ptr, align
+}
+
+impl<'a> Storage<'a> {
+  pub fn new(size: usize, align: Option<usize>) -> Result<Storage<'static>> {
+    Ok(Storage::Owned(Allocation::new(size, align)?))
+  }
+
+  pub fn as_mut_ptr(&self) -> *mut u8 {
+    match self {
+      Storage::Owned(alloc) => alloc.as_mut_ptr(),
+      Storage::View(slice, _) => slice.as_ptr() as *mut u8,
+    }
+  }
+
+  pub fn size(&self) -> usize {
+    match self {
+      Storage::Owned(alloc) => alloc.size(),
+      Storage::View(slice, _) => slice.len(),
+    }
+  }
+
+  pub fn align(&self) -> usize {
+    match self {
+      Storage::Owned(alloc) => alloc.align(),
+      Storage::View(_, align) => *align,
+    }
+  }
+
+  pub fn as_ptr(&self) -> *const u8 {
+    self.as_mut_ptr() as *const _
+  }
+
+  pub fn view(&self) -> Storage<'a> {
+    match self {
+      Storage::Owned(alloc) => Storage::View(
+        unsafe { slice::from_raw_parts_mut(alloc.as_mut_ptr(), self.size()) },
+        self.align(),
+      ),
+      Storage::View(slice, _) => Storage::View(
+        unsafe { slice::from_raw_parts_mut(self.as_mut_ptr(), slice.len()) },
+        self.align(),
+      ),
+    }
+  }
+
+  pub fn is_owned(&self) -> bool {
+    match self {
+      Storage::Owned(_) => true,
+      _ => false,
+    }
+  }
+
+  pub fn to_owned(&self) -> Storage<'static> {
+    let s = Storage::new(self.size(), Some(self.align())).unwrap();
+    unsafe {
+      s.as_mut_ptr()
+        .copy_from_nonoverlapping(self.as_ptr(), self.size())
+    }
+    s
+  }
+}
+
+impl<'a, T> From<&'a [T]> for Storage<'a> {
+  fn from(data: &'a [T]) -> Self {
+    let data = unsafe {
+      slice::from_raw_parts_mut(
+        data.as_ptr() as *const u8 as *mut u8,
+        data.len() * mem::size_of::<T>() as usize,
+      )
+    };
+    Storage::View(data, mem::align_of::<T>())
+  }
+}
+
+#[derive(PartialEq)]
+pub struct Tensor<'a> {
+  pub(super) data: Storage<'a>,
+  pub(super) ctx: TVMContext,
+  pub(super) dtype: DataType,
+  pub(super) shape: Vec<usize>,
+  pub(super) strides: Option<Vec<usize>>,
+  pub(super) byte_offset: isize,
+  pub(super) numel: usize,
+  pub(super) dshape: Vec<i64>,
+}
+
+impl<'a> Tensor<'a> {
+  pub fn shape(&self) -> Vec<usize> {
```

Comment:
Inlinable with `#[inline]` and the like methods.

Reply:
Partially agree! while it's debatable, my reference is bluss's ndarray and the like libs he's wrote with careful inlining. I'm not familiar with wasm though.

Reply:
I'm not sure that it's a debate: LTO is unconditionally better. The ndarray crate is relatively old--perhaps bluss was using best bractices from a time when the Rust linker was less sophisticated.

References:

* https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust
* https://stackoverflow.com/questions/1759300/when-should-i-write-the-keyword-inline-for-a-function-method
* https://stackoverflow.com/questions/7046547/link-time-optimization-and-inline
* https://stackoverflow.com/questions/5948703/difference-in-inlining-functions-by-compiler-or-linker

Reply:
Thanks for the links! I'll look into them. The major source of (anything in fact and) inlining is in std, [for example](https://doc.rust-lang.org/src/core/option.rs.html#191). I think, if in doubt in general, then skip it for now.